tests/TransactionTest.php

```diff
@@ -0,0 +1,430 @@
+
+        $this->assertTrue(DB::collection('users')->where('name', 'klinson')->where('age', 20)->exists());
+    }
+
+    public function testQuery()
+    {
+        /** rollback test */
+        DB::beginTransaction();
+        $count = DB::collection('users')->count();
+        $this->assertEquals(0, $count);
+        DB::collection('users')->insert(['name' => 'klinson', 'age' => 20, 'title' => 'admin']);
+        $count = DB::collection('users')->count();
+        $this->assertEquals(1, $count);
+        DB::rollBack();
+
+        $count = DB::collection('users')->count();
+        $this->assertEquals(0, $count);
+
+        /** commit test */
+        DB::beginTransaction();
+        $count = DB::collection('users')->count();
+        $this->assertEquals(0, $count);
+        DB::collection('users')->insert(['name' => 'klinson', 'age' => 20, 'title' => 'admin']);
+        $count = DB::collection('users')->count();
+        $this->assertEquals(1, $count);
+        DB::commit();
+
+        $count = DB::collection('users')->count();
+        $this->assertEquals(1, $count);
+    }
+
+    public function testTransaction(): void
+    {
+        User::create(['name' => 'klinson', 'age' => 20, 'title' => 'admin']);
+
+        DB::transaction(function (): void {
+            User::create(['name' => 'alcaeus', 'age' => 38, 'title' => 'admin']);
+            User::where(['name' => 'klinson'])->update(['age' => 21]);
+        });
+
+        $count = User::count();
+        $this->assertEquals(2, $count);
+
+        $this->assertTrue(User::where('alcaeus')->exists());
+        $this->assertTrue(User::where(['name' => 'klinson'])->where('age', 21)->exists());
+    }
+
+    public function testTransactionRepeatsOnTransientFailure(): void
+    {
+        User::create(['name' => 'klinson', 'age' => 20, 'title' => 'admin']);
+
+        $timesRun = 0;
+
+        DB::transaction(function () use (&$timesRun): void {
+            User::where(['name' => 'klinson'])->update(['age' => 21]);
+
+            // Update user during transaction to simulate a simultaneous update
+            if ($timesRun == 0) {
+                DB::getCollection('users')->updateOne(['name' => 'klinson'], ['$set' => ['age' => 22]]);
+            }
+
+            $timesRun++;
+        }, 2);
+
+        $this->assertSame(2, $timesRun);
+        $this->assertTrue(User::where(['name' => 'klinson'])->where('age', 21)->exists());
+    }
+
+    public function testTransactionRespectsRepetitionLimit(): void
+    {
+        $klinson = User::create(['name' => 'klinson', 'age' => 20, 'title' => 'admin']);
+
+        $timesRun = 0;
+
+        DB::transaction(function () use (&$timesRun): void {
+            User::where(['name' => 'klinson'])->update(['age' => 21]);
+
+            // Update user during transaction to simulate a simultaneous update
+            DB::getCollection('users')->updateOne(['name' => 'klinson'], ['$inc' => ['age' => 2]]);
+
+            $timesRun++;
+        }, 2);
```

Comment:
I noted you said this [fails in CI](https://github.com/jenssegers/laravel-mongodb/actions/runs/3461141290/jobs/5778375858#step:11:23), but I'm having trouble following how this is executed. I understand that `DB::transaction()` uses a facade and is going to create an active session, but which of the operations within this callback end up using the same session (i.e. updating _within_ the transaction)?

Reply:
The first method call in the transactional closure updates the `user` collection within the transaction. The second call fetches the raw collection class, bypassing the internal transaction logic.

This is the same logic I'm using in the test above this one, with the small difference that in that test, the conflicting update is only done on the first execution of the callback, which means that the transaction will commit successfully after the second callback execution. In this test, I want to specifically cause the transaction to fail every time the callback is run to test whether the repetition limit is observed. This works fine locally but for some reason fails on CI.

Reply:
Are you running a single node replica set with the same server version (I don't think patch is relevant) locally?

Reply:
This test was now fixed 🎉